veracioux/flameshot-dockerfiles
# A new way to manage the dockerfiles

@ZetaoYang

Hi, I've noticed that this repository has a lot of Dockerfiles that are very similar and often almost identical. I created a small bash utility called [shdocker](https://github.com/veracioux/shdocker) that allows you to generate dockerfiles dynamically based on environment variables. For example, use the same file (that looks much like a regular Dockerfile) to generate different dockerfiles depending on the base image or other parameters.

I used this repository as a proof of concept to myself, so I thought it would be appropriate if I shared the concept with you, since I've already done the work on this repo.

So here's what I've done.

#### The end result

I added some files starting with an underscore that are so-called shDockerfiles which shdocker uses to generate Dockerfiles from. For example, the file `_alpine` is used to generate the Dockerfiles that are currently under `alpine/`. Same goes for the dockerfiles for other distros. Ubuntu and Debian are a bit different: since they are very similar, the same shDockerfile (`_debian_ubuntu`) is used to generate both the `debian/` and `ubuntu/` dockerfiles.

The generation process is started by simply running `make`.

Note that the files are simply bash scripts. But they are completely secure, since a docker container is spawned to run them.

#### The changes, step by step

The individual commit messages and/or the changes themselves should be fairly easy to follow. Here I'll just give a birds-eye view.

1. The first step was to rewrite the existing files so that their functionality is not changed, but that they better match what will be the output of shdocker, so that diffs of the following commits are easier to digest.
2. Then I removed the existing Dockerfiles, because they are no longer necessary
3. Then I added a shDockerfile for each distro, each in a separate commit.
4. Then I noticed that `debian` and `ubuntu` are quite similar so I merged `_debian` and `_ubuntu` into a common file: `_debian_ubuntu`.

5. I extracted some common content that appears in (almost) all dockerfiles into a `_header` file, and sourced it from each shDockerfile. That is except for `_centos` because there is a differentiation between `centos:6` and the other versions, which may even be unnecessary (I left a TODO there for you to review).

#### How to review the changes

This is the way I planned for you to test if the generated Dockerfiles are consistent with the existing ones:

- Checkout the HEAD commit (e707d90)
- Run `make` to generate the dockerfiles
- Run `git add .` (necessary for the next step)
- Run `git diff b07d4b8 --cached -- */`
- Verify that the difference between the existing Dockerfiles (rewritten for structure without a change in functionality) and the corresponding Dockerfiles generated by shdocker is not significant in terms of functionality.
- Build the dockerfiles and verify that they work properly

#### Afterword

Thanks for your patience. I hope this will be useful for the project. But even if it isn't, I hope you'll find a use for it in other projects.

**NOTE:** I don't really understand the build process of this repo so I can't test these changes myself. All I could do is verify the changes syntactically.

**NOTE:** This is a proof of concept and the changes are not final. There are some todos that I'd like you to review (just grep the files for TODO).

**NOTE:** shdocker is fairly extensively documented in the repo's README and its manpage.
